File: frontend/src/lib/api/strategyGuard.ts
/**
 * Strategy AST-guard client — backs the AI Drop Zone "Scan" pre-flight.
 *
 * Calls POST /api/strategy-guard/scan, which runs the static AST guard
 * (axiom/sandbox/ast_guard.py) against a strategy file WITHOUT executing it.
 * This is a UX pre-check only; registration enforces the same guard server-side
 * regardless. Kinds mirror the backend's AstReport.Finding.kind exactly.
 */

import { fetchApi } from './core';

export type AstFindingKind =
	| 'forbidden_import'
	| 'dynamic_exec'
	| 'file_too_large'
	| 'too_many_lines'
	| 'syntax_error';

export interface AstFinding {
	kind: AstFindingKind;
	lineno: number;
	col: number;
	message: string;
	node_repr: string;
}

export interface AstReport {
	ok: boolean;
	findings: AstFinding[];
	file_size_bytes: number;
	line_count: number;
}

export async function scanStrategy(path: string): Promise<AstReport> {
	return fetchApi<AstReport>('/strategy-guard/scan', {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify({ path })
	});
}

File: axiom/routers/strategy_guard.py
"""Static AST-guard scan endpoint for strategy files.

Surfaces :func:`Axiom.sandbox.ast_guard.scan_file` over HTTP so the AI Drop
Zone can pre-flight a strategy module BEFORE registering it. This is a *static*
scan only — it never executes the file, spawns a subprocess, or persists a row.

It is a UX convenience, NOT the trust boundary: registration enforces the same
guard server-side (``registry.assert_custom_module_safe`` runs before every
in-process import), so a file that fails this scan is rejected at register time
regardless of whether the caller scanned it first.
"""
from __future__ import annotations

from pathlib import Path
from typing import Any

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field

from axiom.api_security import require_operator_access
from axiom.sandbox.ast_guard import scan_file

router = APIRouter(
    prefix="/api/strategy-guard",
    tags=["strategy-guard"],
    dependencies=[Depends(require_operator_access)],
)


class _ScanBody(BaseModel):
    path: str = Field(..., description="Filesystem path to a .py strategy file")


@router.post("/scan")
def post_scan(body: _ScanBody) -> dict[str, Any]:
    """Run the static AST guard against *path* and return its findings."""
    p = Path(body.path)
    if not p.exists():
        raise HTTPException(
            status_code=404, detail={"error": "path_not_found", "path": str(p)}
        )
    if not p.is_file():
        raise HTTPException(
            status_code=400, detail={"error": "not_a_file", "path": str(p)}
        )

    report = scan_file(p)
    return {
        "ok": report.ok,
        "findings": [
            {
                "kind": f.kind,
                "lineno": f.lineno,
                "col": f.col,
                "message": f.message,
                "node_repr": f.node_repr,
            }
            for f in report.findings
        ],
        "file_size_bytes": report.file_size_bytes,
        "line_count": report.line_count,
    }
